src/move/emojicoin_arena/sources/emojicoin_arena.move

```diff
@@ -0,0 +1,713 @@
+        randomness::u64_range(0, n_markets) + 1
+    }
+
+    inline fun register_melee(
+        registry_ref_mut: &mut Registry,
+        n_melees_before_registration: u64,
+        sorted_unique_market_ids: vector<u64>
+    ) {
+        let melee_id = n_melees_before_registration + 1;
+        registry_ref_mut.melees_by_id.add(
+            melee_id,
+            Melee {
+                melee_id,
+                market_metadatas: sorted_unique_market_ids.map_ref(|market_id_ref| {
+                    option::destroy_some(
+                        emojicoin_dot_fun::market_metadata_by_market_id(*market_id_ref)
+                    )
+                }),
+                start_time: last_period_boundary(
+                    timestamp::now_microseconds(), registry_ref_mut.new_melee_duration
+                ),
+                duration: registry_ref_mut.new_melee_duration,
+                max_match_percentage: registry_ref_mut.new_melee_max_match_percentage,
+                max_match_amount: registry_ref_mut.new_melee_max_match_amount,
+                available_rewards: registry_ref_mut.new_melee_available_rewards,
+                all_entrants: smart_table::new(),
+                active_entrants: smart_table::new(),
+                exited_entrants: smart_table::new(),
+                locked_in_entrants: smart_table::new(),
+                n_melee_swaps: aggregator_v2::create_unbounded_aggregator(),
+                melee_swaps_volume: aggregator_v2::create_unbounded_aggregator(),
+                emojicoin_0_locked: aggregator_v2::create_unbounded_aggregator(),
+                emojicoin_1_locked: aggregator_v2::create_unbounded_aggregator()
+            }
+        );
+        registry_ref_mut.melee_ids_by_market_ids.add(sorted_unique_market_ids, melee_id);
+    }
+
+    inline fun remove_if_contains(
+        map_ref_mut: &mut SmartTable<u64, Nil>, key: u64
+    ) {
+        if (map_ref_mut.contains(key)) {
+            map_ref_mut.remove(key);
+        }
+    }
+
+    inline fun sort_unique_market_ids(market_id_0: u64, market_id_1: u64): vector<u64> {
+        if (market_id_0 < market_id_1) {
+            vector[market_id_0, market_id_1]
+        } else {
+            vector[market_id_1, market_id_0]
+        }
+    }
+
+    inline fun swap_with_stats<Emojicoin, LP>(
+        swapper: &signer,
+        swapper_address: address,
+        market_address: address,
+        input_amount: u64,
+        is_sell: bool,
+        integrator_fee_rate_bps: u8
+    ): (u64, u64) {
+        let simulated_swap =
+            emojicoin_dot_fun::simulate_swap<Emojicoin, LP>(
+                swapper_address,
+                market_address,
+                input_amount,
+                is_sell,
+                @integrator,
+                integrator_fee_rate_bps
+            );
+        let (_, _, _, _, _, _, _, _, net_proceeds, _, quote_volume, _, _, _, _, _, _, _) =
+            emojicoin_dot_fun::unpack_swap(simulated_swap);
```

Comment:
Noting here these tuple unpacks are ripe for errors/typos, although I trust that you did it correctly ;) I think it's good for a double check just in case, though, or maybe a really simple helper function with a simple unit test

It would be nice if there were a concept of public/private fields though so this wouldn't be necessary...

Reply:
> public/private fields

In the absence of such a paradigm I unfortunately think this is the best we'll get

And I started writing such a test, then realized at a certain point you just end up writing the same unpack syntax inside the test itself...

I think that due to the randomness APIs etc. and issues with potentially testing everything in unit tests a lot of this might have to be e2e...

Reply:
Lol yes that's true, it really does seem like going in circles